File: VoidSchach960/src/main/java/player/ki/concurrent/MultiThreadStrategy.java
package player.ki.concurrent;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicInteger;

import helper.Move;
import player.ki.AbstractComputerPlayerUI;
import player.ki.DynamicEvaluation;
import board.ChessGameInterface;

class MultiThreadStrategy extends AbstractConcurrencyStrategy
{
  final private DelayedExecutorService delayedExecutorService;
  
	MultiThreadStrategy(AbstractComputerPlayerUI ui, int numberOfCores)
	{
		super(ui);
		
		delayedExecutorService = new DelayedExecutorService(numberOfCores);
	}
	

	@Override
	public SortedSet<EvaluatedMove> evaluatePossibleMoves(final ChessGameInterface game, final DynamicEvaluation dynamicEvaluation)
	{
	  //as long as the first parameter is 0 and the second one is bigger
	  //the progress bar will always show correctly 0% (so '1' as second parameter is fine)
	  showProgress(0, 1);
	  
		final List<Move> possibleMoves = getPossibleMoves(game);
		assert !possibleMoves.isEmpty() : "no moves were possible and therefore evaluatable";
		
		final int totalNumberOfMoves = possibleMoves.size();
		final AtomicInteger numberOfFinnishedEvaluations = new AtomicInteger(0);
		
		final Iterator<ChessGameInterface> gameInstances = game.copyGame(totalNumberOfMoves).iterator();
		
		final LinkedList<Callable<EvaluatedMove>> movesToEvaluate = new LinkedList<Callable<EvaluatedMove>>();
		for(Move move: possibleMoves) {
		  movesToEvaluate.add(
        new MoveEvaluationCallable(
            gameInstances.next(), move, dynamicEvaluation, numberOfFinnishedEvaluations
        )
      );
    }
		delayedExecutorService.setCallablesToExecute(movesToEvaluate);
		
		long time = System.currentTimeMillis();
		
		final List<Future<EvaluatedMove>> evaluationFutures = blockAndShowProgressUntillEvaluationIsDone(numberOfFinnishedEvaluations, totalNumberOfMoves);
		
    final SortedSet<EvaluatedMove> result = new TreeSet<EvaluatedMove>();
    for (Future<EvaluatedMove> evaluatedMoveFuture : evaluationFutures) {
      try {
        //get() shouldn't need to block but it shouldn't be a problem is if it does
        final EvaluatedMove evaluatedMove = evaluatedMoveFuture.get();
        if(evaluatedMove==null) {
          //only possible in case of exception
          continue;
        }
        result.add(evaluatedMove);
      } catch (Exception e) {/*shouldn't happen*/}
    }
    
    long duration = time - System.currentTimeMillis();
    System.out.println("time:"+duration);
    
    assert !result.isEmpty() : "no evaluation of a possible moves was successfull";
    
    return result;
	}
	
	private List<Future<EvaluatedMove>> blockAndShowProgressUntillEvaluationIsDone(final AtomicInteger numberOfFinnishedEvaluations, final int totalNumberOfMoves)
	{
	  final List<Future<EvaluatedMove>> evaluationFutures = new ArrayList<Future<EvaluatedMove>>(totalNumberOfMoves);
	  
	  delayedExecutorService.executeOneCallableForEachThread(evaluationFutures);
	  	  
	  int lastNumberOfFinnishedEvaluations = 0;
	  boolean notAllEvaluationsDone = true;
	  
	  while(notAllEvaluationsDone) {
	    //sleep for a second
	    sleep(300);
	    //than check the number of finnished Evaluations 
	    
	    final int nowNumberOfFinnishedEvaluations = numberOfFinnishedEvaluations.get();
	    if(nowNumberOfFinnishedEvaluations>lastNumberOfFinnishedEvaluations) {
	      //more evaluations have been finnished
	      //so update the lastNumberOfFinnishedEvaluations for the next iteration
	      final int freeThreads = nowNumberOfFinnishedEvaluations - lastNumberOfFinnishedEvaluations;
	      lastNumberOfFinnishedEvaluations = nowNumberOfFinnishedEvaluations;
	      //and update the progress bar
     
	      showProgress(nowNumberOfFinnishedEvaluations, totalNumberOfMoves);
	      
        delayedExecutorService.executeCallables(freeThreads, evaluationFutures);
	    }
	    //check if all evaluations are done
	    notAllEvaluationsDone = nowNumberOfFinnishedEvaluations<totalNumberOfMoves; 
	  }
	  
	  return evaluationFutures;
	}
	
	private void sleep(final int numberOfMilliseconds)
	{
	  try{
	    Thread.sleep(numberOfMilliseconds);
	  } catch(InterruptedException e) {}
	}
	
//	private synchronized void moveEvaluated()
//	{
//		numberOfEvaluatedMoves++;
//		
//		showProgress();
//	}
//	
//	private synchronized void addException(Move move, Exception e)
//	{		
//		//TODO make better message
//		exceptions.add(move.toString()+" caused "+e.toString());
//		
//		moveEvaluated();
//	}
//	
//	private void showProgress()
//	{
//		showProgress(numberOfEvaluatedMoves, totalNumberOfMoves);
//	}
//	
//	private void assertNoExceptionsWhileEvaluatingGame(ChessGameInterface game)
//	{
//		//TODO better exception handling (e.g. incorporate game as String)
//		if( !exceptions.isEmpty() ) {
//			StringBuilder sb = new StringBuilder();
//			for(String msg: exceptions) {
//				sb.append(msg).append("; \n");
//			}
//			
//			throw new RuntimeException("exception(s) while computing moves: \n"+sb.toString());
//		}
//	}
	
	private static class MoveEvaluationCallable implements Callable<EvaluatedMove>
	{
	  final AtomicInteger numberOfFinnishedEvaluations;
		private final DynamicEvaluation dynamicEvaluation;
		private final ChessGameInterface game;
		private final Move move; 
		
		MoveEvaluationCallable(ChessGameInterface game, Move move, DynamicEvaluation dynamicEvaluation, final AtomicInteger numberOfFinnishedEvaluations)
		{
		  this.numberOfFinnishedEvaluations = numberOfFinnishedEvaluations;
			this.dynamicEvaluation = dynamicEvaluation;
			this.game = game;
			this.move = move;
		}

		public EvaluatedMove call() throws Exception {
			try {
				final float value = dynamicEvaluation.evaluateMove(game, move);
				EvaluatedMove evaluatedMove = new EvaluatedMove(move, value);

				return evaluatedMove;
			} catch (Exception e) {
			  //print out the error
			  e.printStackTrace(System.err);
			  return null;
			}finally{
			  //this Callable is finished so increment this counter
			  //so that the progress can be examined
				numberOfFinnishedEvaluations.incrementAndGet();
			}
		}
	}
	
	private static class DelayedExecutorService
	{
	  private final ExecutorService executorService;
	  private final int numberOfThreads;
	  
	  private LinkedList<Callable<EvaluatedMove>> callablesToExecute;
	  
	  DelayedExecutorService(int numberOfCores)
	  {	    
	    numberOfThreads = numberOfCores;
	    executorService = Executors.newFixedThreadPool(numberOfCores);
	  }
	  
	  public void setCallablesToExecute(LinkedList<Callable<EvaluatedMove>> callablesToExecute)
	  {
	    this.callablesToExecute = callablesToExecute;
	  }
	  
	  public void executeOneCallableForEachThread(List<Future<EvaluatedMove>> resultFutures)
	  {
	    executeCallables(numberOfThreads, resultFutures);
	  }
	  
	  public void executeCallables(final int numberOfCallablesToExecute, List<Future<EvaluatedMove>> resultFutures)
    {      
      for(int i=0;i<numberOfThreads;i++) {
        if(callablesToExecute.isEmpty()) {
          break;
        }
        resultFutures.add(
          executorService.submit(
            callablesToExecute.removeFirst()
          )
        );
      }
    }
	}
}

File: module-engine-internal/src/main/kotlin/voidchess/engine/player/ki/concurrent/MultiThreadStrategy.kt
package voidchess.engine.player.ki.concurrent

import voidchess.common.board.move.Move
import voidchess.common.player.ki.ProgressCallback
import voidchess.common.player.ki.evaluation.EvaluatedMove
import voidchess.common.player.ki.evaluation.Evaluation
import voidchess.engine.board.EngineChessGame
import voidchess.engine.player.ki.evaluation.EvaluatingMinMax
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicReference
import kotlin.collections.ArrayList


internal class MultiThreadStrategy(
    private val numberOfThreads: Int,
    progressCallback: ProgressCallback
) : ConcurrencyStrategy(progressCallback) {

    private val executorService: ExecutorService = Executors.newFixedThreadPool(numberOfThreads)

    override fun shutdown() {
        executorService.shutdown()
    }


    override fun evaluateMoves(
        game: EngineChessGame,
        movesToEvaluate: Collection<Move>,
        evaluatingMinMax: EvaluatingMinMax,
        numericEvalOkRadius: Double
    ): MutableList<EvaluatedMove> {

        val currentMaxEvaluationRef = AtomicReference<Evaluation?>(null)
        val callablesToEvaluate =
            getEvaluableMoves(game, movesToEvaluate, evaluatingMinMax, currentMaxEvaluationRef, numericEvalOkRadius)

        val result = try {
            evaluate(callablesToEvaluate)
        } catch (e: Exception) {
            e.printStackTrace()
            LinkedList<EvaluatedMove>()
        }

        assert(result.isNotEmpty()) { "no evaluation of a possible moves was successful" }

        return result
    }

    @Throws(InterruptedException::class, ExecutionException::class)
    private fun evaluate(movesToEvaluate: LinkedList<Callable<EvaluatedMove>>): MutableList<EvaluatedMove> {
        val totalNumberOfMoves = movesToEvaluate.size
        val result = ArrayList<EvaluatedMove>(totalNumberOfMoves)

        val ecs = ExecutorCompletionService<EvaluatedMove>(executorService)
        submitCallables(movesToEvaluate, ecs, numberOfThreads)

        for (i in 0 until totalNumberOfMoves) {
            //wait for an evaluation to be finished
            val evaluatedMove = ecs.take().get()
            //show the progress
            progressCallback(i, totalNumberOfMoves)
            //add a new move to be evaluated to the queue (if some are left)
            submitCallables(movesToEvaluate, ecs, 1)
            //add this evaluation to result set
            if (evaluatedMove != null) {
                result.add(evaluatedMove)
            }
        }
        progressCallback(totalNumberOfMoves, totalNumberOfMoves)

        return result
    }

    private fun getEvaluableMoves(
        game: EngineChessGame,
        movesToEvaluate: Collection<Move>,
        evaluatingMinMax: EvaluatingMinMax,
        currentMaxEvaluationRef: AtomicReference<Evaluation?>,
        numericEvalOkRadius: Double
    ): LinkedList<Callable<EvaluatedMove>> {
        assert(movesToEvaluate.isNotEmpty()) { "no moves were possible and therefore evaluable" }

        val totalNumberOfMoves = movesToEvaluate.size

        // TODO instead of generating one game per move, generate one game per thread
        val gameInstances = game.copyGame(totalNumberOfMoves).iterator()

        val callablesToEvaluate = LinkedList<Callable<EvaluatedMove>>()
        for (move in movesToEvaluate) {
            callablesToEvaluate.add(
                MoveEvaluationCallable(
                    gameInstances.next(), move, evaluatingMinMax, currentMaxEvaluationRef, numericEvalOkRadius
                )
            )
        }

        return callablesToEvaluate
    }

    private fun submitCallables(movesToEvaluate: LinkedList<Callable<EvaluatedMove>>, completionService: CompletionService<EvaluatedMove>, numberOfMovesToSubmit: Int) {
        for (i in 0 until numberOfMovesToSubmit) {
            if (movesToEvaluate.isEmpty()) {
                break
            }
            completionService.submit(movesToEvaluate.removeFirst())
        }
    }

    private class MoveEvaluationCallable internal constructor(
        private val game: EngineChessGame,
        private val move: Move,
        private val evaluatingMinMax: EvaluatingMinMax,
        private val currentMaxEvaluationRef: AtomicReference<Evaluation?>,
        private val numericEvalOkRadius: Double
    ) : Callable<EvaluatedMove> {

        @Throws(Exception::class)
        override fun call(): EvaluatedMove? {
            return try {
                val currentMaxEvaluation = currentMaxEvaluationRef.get()
                val currentOkEval: Evaluation? = getOkEval(currentMaxEvaluation, numericEvalOkRadius)

                val latestEvaluation = evaluatingMinMax.evaluateMove(game, move, currentOkEval)

                if (currentOkEval == null || latestEvaluation > currentOkEval) {
                    currentMaxEvaluationRef.updateAndGet { latestMaxEvaluation: Evaluation? ->
                        if (latestMaxEvaluation == null || latestEvaluation > latestMaxEvaluation) {
                            latestEvaluation
                        } else {
                            latestMaxEvaluation
                        }
                    }
                    EvaluatedMove(move, latestEvaluation)
                } else {
                    null
                }
            } catch (e: Exception) {
                //print out the error
                e.printStackTrace()
                null
            }
        }
    }
}
